### PR TITLE
AWS X-Ray

### PR DESCRIPTION
**Pull request by terrycain**
Started work on X-Ray cuz it looked cool and #1006 wanted it. Now I regret it, its like one of AWS's secrets giving out little documentation. 

Will probably need actual xray data from the community to get the data structures and logic locked down but I have make a good crack at it which will be a decent starting point. Also helps AWS are using a different API structure yet again :smiley:.

Want to implement the last 3 endpoints, to return correct but empty data, least that way boto wont blow up.

TODO:
- [x] BatchGetTraces
- [x] GetServiceGraph
- [x] GetTraceGraph
- [x] GetTraceSummaries
- [x] PutTelemetryRecords
- [x] PutTraceSegments

### CODE REVIEW

**Comment by terrycain** [comment]
All the methods are mocked. GetServiceGraph, GetTraceGraph just return acceptable json so boto doesnt go mental.

Finished for now :smile:

File: moto/xray/responses.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    def request_params(self):
+        try:
+            return json.loads(self.body)
+        except ValueError:
+            return {}
+
+    def _get_param(self, param, default=None):
+        return self.request_params.get(param, default)
+
+    def _get_action(self):
```
Comment (JackDanger) [inline moto/xray/responses.py]: Nice work pulling out `_get_action`

**Comment by JackDanger** [comment]
This is a fantastic start to XRay, @terrycain - you're on a damned roll

File: moto/xray/responses.py (hunk 1 above)
Comment (terrycain) [reply moto/xray/responses.py]: Yeah I had just reimplemented the call function but then it was bugging me any time I looked at that class :-)

**Comment by terrycain** [comment]
Yeah, looking at X-ray and some of the microservice stuff I do at work, I reckon it'll be useful soon so we should see a decent implementation in the not so distant future.

Can we add a UDP mode to moto? or branch out to another repo to either mock the aws xray daemon or spin up the aws xray daemon in test mode, as thats the real thing that needs mocking

**Comment by JackDanger** [comment]
@terrycain That's an interesting idea. I don't know why Moto server can't bind to a UDP socket. If AWS does it it seems reasonable to have Moto server do it.

(this PR is now released as Moto 1.1.14)

**Pull request merged**